### PR TITLE
Personify SSO adapter

### PR DESCRIPTION
**Pull request by rotator**
@anpolimus please, review.
Steps to test:
1. Personify module have to be fixed as well.
2. Setup Personify configs in settings.php according to [README.md](https://github.com/fivejars/openy_gated_content/blob/5d3d4b85b5470fa5ba88d4e164d6bb9064f11067/modules/openy_gc_auth/modules/openy_gc_auth_personify/README.md)

### CODE REVIEW

File: js/gated-content/src/store/modules/auth/personify.js [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+  actions: {
+    async personifyAuthorize(context) {
+      return client.post(context.getters.getPersonifyConfig.api_login_check)
+        .then((r) => {
+          // Check user data.
+          if (r.data.user && r.data.user.email) {
+            context.dispatch('authorize', r.data.user);
+          } else {
+            // Get Personify login URL and redirect to login page.
+            client({
```
Comment (anpolimus) [inline js/gated-content/src/store/modules/auth/personify.js:16]: We can construct link on backend side.
There is a method at gc plugin, called getDataForApp().

File: modules/openy_gc_auth/modules/openy_gc_auth_personify/src/Controller/PersonifyAuthController.php [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+   * @param \Symfony\Component\HttpFoundation\RequestStack $requestStack
+   *   Request stack.
+   * @param \Drupal\Core\Logger\LoggerChannelFactory $loggerChannelFactory
+   *   Logger factory.
+   */
+  public function __construct(
+    PersonifySSO $personifySSO,
+    PersonifyClient $personifyClient,
+    ConfigFactoryInterface $configFactory,
+    RequestStack $requestStack,
```
Comment (anpolimus) [inline modules/openy_gc_auth/modules/openy_gc_auth_personify/src/Controller/PersonifyAuthController.php:74]: There is no need to add this service. Every controller method receive's Request object as first parameter by default.

File: modules/openy_gc_auth/modules/openy_gc_auth_personify/src/Controller/PersonifyAuthController.php [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+
+    return new JsonResponse($loginUrl);
+  }
+
+  /**
+   * Set cookies to authenticate user based on response from Personify.
+   *
+   * @throws \GuzzleHttp\Exception\GuzzleException
+   */
+  public function auth() {
```
Comment (anpolimus) [inline modules/openy_gc_auth/modules/openy_gc_auth_personify/src/Controller/PersonifyAuthController.php:138]:
```suggestion
  public function auth(Request $request) {
```

File: modules/openy_gc_auth/modules/openy_gc_auth_personify/src/Plugin/GCIdentityProvider/Personify.php [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+
+    $form['endpoints'] = [
+      '#type' => 'details',
+      '#title' => t('Custom Personify endpoints'),
+      '#collapsible' => TRUE,
+      '#open' => FALSE,
+      '#description' => $this->t('Change these values only in case you have custom endpoints for every URL.'),
+    ];
+    $form['endpoints']['api_login_check'] = [
+      '#title' => $this->t('Login check endpoint'),
```
Comment (anpolimus) [inline modules/openy_gc_auth/modules/openy_gc_auth_personify/src/Plugin/GCIdentityProvider/Personify.php:66]: There is no need to set routes that you've created yourself.
These routes were created here: https://github.com/fivejars/openy_gated_content/pull/31/files#diff-6bbd677965587a158ae13f2d460bd247 and there is no need to ask administrator to enter them again.
All that we need here is to save Virtual Ymca page, where you will redirect user at the end.

File: modules/openy_gc_auth/modules/openy_gc_auth_personify/config/install/openy_gc_auth.settings.yml [hunk 5]
```diff
@@ -0,0 +1 @@
+active_provider: personify
```
Comment (anpolimus) [inline modules/openy_gc_auth/modules/openy_gc_auth_personify/config/install/openy_gc_auth.settings.yml:1]: Why do we need this config here?

**Comment by anpolimus** [comment]
@rotator 
Remaining changes: 
1. Please implement Spinner for your auth component like here: https://github.com/fivejars/openy_gated_content/pull/37
2. Please rebuild logic for check of the logged in status like here: https://github.com/fivejars/openy_gated_content/pull/36

File: modules/openy_gc_auth/modules/openy_gc_auth_personify/src/Plugin/GCIdentityProvider/Personify.php [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+   * {@inheritdoc}
+   */
+  public function buildConfigurationForm(array $form, FormStateInterface $form_state) {
+    $config = $this->getConfiguration();
+    $form = parent::buildConfigurationForm($form, $form_state);
+
+    $personifyConfigForm = Link::createFromRoute($this->t('Personify configuration form'),
+      'personify.settings', [], ['target' => '_blank'])->toString();
+    $readMeUrl = Link::fromTextAndUrl($this->t('README file'),
+      Url::fromUserInput('/' . drupal_get_path('module', 'openy_gc_auth_personify') . '/README.md',
```
Comment (podarok) [inline modules/openy_gc_auth/modules/openy_gc_auth_personify/src/Plugin/GCIdentityProvider/Personify.php:105]: `*.md` file extension should be enabled on the server level for downloading.

Comment (rotator) [reply modules/openy_gc_auth/modules/openy_gc_auth_personify/src/Plugin/GCIdentityProvider/Personify.php:105]: I can change it to *.txt then or completely remove the link.
What will be more suitable and common for the OpenY?

Comment (podarok) [reply modules/openy_gc_auth/modules/openy_gc_auth_personify/src/Plugin/GCIdentityProvider/Personify.php:105]: we could be using hook_help() in order to not introduce environmental issues
*.txt could work too, but not all webhost setups allowing these as well

**Comment by AndreyMaximov** [comment]
retest

**Pull request merged**